### PR TITLE
Type alias 3

### PR DESCRIPTION
**Pull request by dedbox**
Continuing the good work started in #505 .

This PR migrates the positive changes from PR #505 while eliding spurious edits and undesirable extensions to internal APIs, such as persistent modification of elaboration contexts in the statics.

### CODE REVIEW

**Comment by dedbox** [comment]
I've also been keeping this branch up to date with #558 for code exploration and some unit testing.

File: Makefile [hunk 1]
```diff
@@ -57,7 +57,10 @@ repl:
 
 test:
 	dune build @src/fmt --auto-promote || true
-	dune runtest || true
+	dune exec src/hazeltest/hazeltest.exe -- --regression-dir src/hazeltest/regressions
```
Comment (cyrus-) [inline Makefile]: should this be part of this PR, or should we pull it out and refactor with the tezt pr?

File: docs/Advanced-Logging.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
+The js_of_ocaml devs recommended `logs` for performance and flexibility:
```
Comment (cyrus-) [inline docs/Advanced-Logging.md]: Make this its own PR

File: opam.export [hunk 3]
```diff
@@ -1,5 +1,4 @@
 opam-version: "2.0"
```
Comment (cyrus-) [inline opam.export]: why is this file changed relative to dev?

File: src/hazelcore/UHExp.rei [hunk 4]
```diff
@@ -106,3 +109,5 @@ let associate: seq => Skel.t(Operators_Exp.t);
 let mk_OpSeq: OpSeq.seq(operand, operator) => OpSeq.t(operand, operator);
 
 let is_complete: t => bool;
+
+/* let fix_indices: t => t; */
```
Comment (cyrus-) [inline src/hazelcore/UHExp.rei]: remove

File: src/hazelcore/UHPat.re [hunk 5]
```diff
@@ -204,3 +209,9 @@ and is_complete_operand = (operand: 'operand): bool => {
   | Inj(NotInHole, _, body) => is_complete(body)
   };
 };
+
+/* let defines_var = (OpSeq(skel, seq): t): bool => */
```
Comment (cyrus-) [inline src/hazelcore/UHPat.re]: remove

File: src/hazelcore/Action_Pat.re [hunk 6]
```diff
@@ -1560,25 +1584,39 @@ and ana_perform_operand =
...
-      } else {
-        let (new_zopseq, id_gen) =
-          new_zopseq |> ZPat.mk_inconsistent(id_gen);
-        Succeeded((new_zopseq, ctx, id_gen));
-      };
+    | Succeeded((zann, ctx, id_gen)) =>
+      switch (Elaborator_Typ.syn_elab(ctx, Delta.empty, ZTyp.erase(zann))) {
+      | None => Failed
+      | Some((ty', _, _)) =>
+        switch (UHPat.get_var_name(op)) {
```
Comment (cyrus-) [inline src/hazelcore/Action_Pat.re]: don't understand what's going on here? why are we special casing when the annotatin is on a variable?

File: src/hazelcore/UHPat.re [hunk 7]
```diff
@@ -45,6 +45,11 @@ let floatlit = (~err: ErrStatus.t=NotInHole, f: string) => FloatLit(err, f);
 
 let listnil = (~err: ErrStatus.t=NotInHole, ()) => ListNil(err);
 
+let get_var_name: operand => option(Var.t) =
```
Comment (cyrus-) [inline src/hazelcore/UHPat.re]: see comment in Action_Pat -- not sure what's going on there and why this is necessary

File: src/hazelcore/TPat.re [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  [@deriving sexp]
+  type t =
+    | ReservedKeyword
+    | BuiltinType
+    /* TODO: (eric) rename InvalidName to InvalidText */
+    | InvalidName;
+};
+
+/* type variable pattern hole status */
+module Status = {
```
Comment (cyrus-) [inline src/hazelcore/TPat.re]: call it ErrStatus to be consistent with other sorts

File: src/hazelcore/TPat.re [hunk 9]
```diff
@@ -0,0 +1,50 @@
+open Sexplib.Std;
+
+/* type variable pattern errors */
+module HoleReason = {
+  [@deriving sexp]
+  type t =
+    | ReservedKeyword
+    | BuiltinType
+    /* TODO: (eric) rename InvalidName to InvalidText */
```
Comment (cyrus-) [inline src/hazelcore/TPat.re]: yes, do that renaming :)

File: src/hazelcore/TPat.re [hunk 10]
```diff
@@ -0,0 +1,50 @@
+open Sexplib.Std;
+
+/* type variable pattern errors */
+module HoleReason = {
```
Comment (cyrus-) [inline src/hazelcore/TPat.re]: move as a submodule of ErrStatus module

File: src/hazelcore/TPat.re [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  [@deriving sexp]
+  type t =
+    | NotInHole
+    | InHole(HoleReason.t, MetaVar.t);
+};
+
+[@deriving sexp]
+type t =
+  | EmptyHole
+  | TyVar(Status.t, string);
```
Comment (cyrus-) [inline src/hazelcore/TPat.re]: use TyVar.t instead of string here, and have InvalidText be its own top-level constructor of this type

File: src/hazelcore/TyVar.re [hunk 12]
```diff
@@ -0,0 +1,30 @@
+open Sexplib.Std;
+
+[@deriving sexp]
+type t = string;
+
+let length: t => int = String.length;
+
+let equal: (t, t) => bool = String.equal;
+
```
Comment (cyrus-) [inline src/hazelcore/TyVar.re:9]: add a valid_regex top level binding following Var

File: src/hazelcore/TyVar.re [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+let builtin_type: string => bool =
+  fun
+  | "Int"
+  | "Float"
+  | "Bool" => true
+  | _ => false;
+
+let keyword: string => bool =
+  fun
+  | "let"
```
Comment (cyrus-) [inline src/hazelcore/TyVar.re]: don't duplicate all the keywords here, use the ExpandingKeyword module

File: src/hazelcore/TyVar.re [hunk 14]
```diff
@@ -0,0 +1,30 @@
+open Sexplib.Std;
```
Comment (cyrus-) [inline src/hazelcore/TyVar.re:1]: missing TyVar.rei

File: src/hazelcore/UHTyp.re [hunk 15]
```diff
@@ -5,6 +5,7 @@ type operator = Operators_Typ.t;
 type t = opseq
 and opseq = OpSeq.t(operand, operator)
 and operand =
+  | TyVar(TyVarErrStatus.t, TyVar.t)
```
Comment (cyrus-) [inline src/hazelcore/UHTyp.re:10]: Add InvalidText as a top level entry here to follow UHExp instead of putting it all in TyVar

**Review by cyrus- — CHANGES_REQUESTED** [review]
Could only find one small behavioral bug:
* edit action history panel is still seeing bug where spacing is screwed up due width of "tvar" badge when constructing a tpat

Started reviewing the code, namely just the new syntax introduced in this PR, and some changes are needed -- see code comments.

File: Makefile (hunk 1 above)
Comment (dedbox) [reply Makefile]: This code was inherited directly from the tezt pr, so we can just undo it. There may be some unit tests specific to type aliases / tyvars. If so, we can park them on a separate branch and then later either fold into tezt pr or make a new one.

File: opam.export (hunk 3 above)
Comment (dedbox) [reply opam.export]: Probably for logs or tezt. Any relevant changes should've been inherited from one of those two branches directly and can be undone safely.

File: src/hazelcore/Action_Pat.re (hunk 6 above)
Comment (dedbox) [reply src/hazelcore/Action_Pat.re]: I think these two cases correspond to whether or not we extend the given context. If get_var_name returns Some type, then the pattern part is not something weird (i.e., it's just a pattern variable) and we try to extend the context. Otherwise, we don't try.

**Comment by dedbox** [comment]
> edit action history panel is still seeing bug where spacing is screwed up due width of "tvar" badge when constructing a tpat

I can't reproduce this in latest chrome or firefox. Maybe you used a different browser?

File: Makefile (hunk 1 above)
Comment (dedbox) [reply Makefile]: fixed by b283048cf and 9a8f1049c

File: docs/Advanced-Logging.md (hunk 2 above)
Comment (dedbox) [reply docs/Advanced-Logging.md]: fix at #658 in progress

File: opam.export (hunk 3 above)
Comment (dedbox) [reply opam.export]: fixed by 9a8f1049c

File: src/hazelcore/UHExp.rei (hunk 4 above)
Comment (dedbox) [reply src/hazelcore/UHExp.rei]: fixed by 737afe234

File: src/hazelcore/UHPat.re (hunk 5 above)
Comment (dedbox) [reply src/hazelcore/UHPat.re]: fixed by 737afe234

File: src/hazelcore/TyVar.re (hunk 13 above)
Comment (dedbox) [reply src/hazelcore/TyVar.re]: Roger.

There are two equivalent functions in ExpandingKeyword.re: `of_string` is a simple fun on strings and `mk` is an equivalent series of if statements that calls the various `is_*` predicates.

Just dropping `mk` would be very easy.

**Comment by cyrus-** [comment]
This is Firefox on Linux. Notice misaligned arrows on right.
Screenshot from 2022-07-17 16-48-57

File: src/hazelcore/TPat.re (hunk 11 above)
Comment (dedbox) [reply src/hazelcore/TPat.re]: fixed by e0da4528d

File: src/hazelcore/TyVar.re (hunk 12 above)
Comment (dedbox) [reply src/hazelcore/TyVar.re:9]: fixed by e0da4528d

File: src/hazelcore/TyVar.re (hunk 13 above)
Comment (dedbox) [reply src/hazelcore/TyVar.re]: > don't duplicate all the keywords here, use the ExpandingKeyword module

fixed by e0da4528d

File: src/hazelcore/TyVar.re (hunk 14 above)
Comment (dedbox) [reply src/hazelcore/TyVar.re:1]: fixed by 622a28a03

File: src/hazelcore/UHTyp.re (hunk 15 above)
Comment (dedbox) [reply src/hazelcore/UHTyp.re:10]: fixed by 82b935785 and f344ae350 and 22e9d99e0

**Comment by cyrus-** [comment]
Archiving this. We are exploring some new approaches that are less finicky than de Bruijn indices now, but this might be a useful reference in any case -- see #612.
